[styles] Add ref prop to withTheme

ianschmitz opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui-org/material-ui/blob/master/CONTRIBUTING.md#submitting-a-pull-request).

`withTheme` now forwards refs, but the typing for the `ref` prop is missing in the declaration file.

eps1lon commented on packages/material-ui-styles/src/withTheme/withTheme.d.ts:
```diff
@@ -13,6 +13,7 @@ export function withThemeCreator<Theme = DefaultTheme>(
 export interface WithTheme<Theme = DefaultTheme> {
   theme: Theme;
   innerRef?: React.Ref<any>;
+  ref?: React.Ref<any>;
```
Since this is new and we do rely on ^3.0 I'd rather use `unknown` here. (we probably need to widen this later but let's not start out with legacy any.)

eps1lon commented:
I just realized that this is probably the wrong place. The `WithTheme` interface should be used for props of components decorated with `withTheme`.

This would now imply that you have access inside your component to `props.ref`. However, `ref` (like `key`) will be omitted.


```tsx
interface Props extends WithTheme {}

function Component(props: Props) {
  const {ref} = props; // whoops, this can never happen
}

export default withTheme(Component);
```

Could you add a usage example that would be allowed with this change (i.e. a test) to https://github.com/mui-org/material-ui/blob/2bcc874cf07b81202968f769cb9c2398c7c11311/packages/material-ui-styles/test/styles.spec.tsx#L70?

ianschmitz commented:
Oh shoot sorry I lost track of this. I'll take a look today!

ianschmitz commented:
Alright I've updated it. I think it's what you were after. I've also updated `material-ui-styles/src/withTheme/withTheme.d.ts` to match.

Not super happy with the typings in `material-ui-styles`. Let me know if you're okay with it.

ianschmitz commented:
Also a question. I noticed `innerRef` is on `WithTheme` in `material-ui-styles/src/withTheme/withTheme.d.ts`. If you're happy with the new interface i created (`ThemedComponentProps`) shall i move it in there? Today there's the same issue @eps1lon identified above.

eps1lon commented:
@ianschmitz Perfect, thanks!
